design/open/spark-outputcommitter/committer.md

```diff
@@ -0,0 +1,194 @@
+# Proposal
+
+Add a Hadoop OutputCommitter that uses _existing_ lakeFS operations for
+**atomic** commits that are efficient and safely concurrent.
+
+## Terminology
+
+lakeFS and Hadoop both use the term "commit", but with different meanings.
+In this document:
+  * A [_lakeFS commit_][lakefs-commit] is a revision entry in the lakeFS
+    repository log.
+  * A [_Hadoop OutputCommitter commit_][hadoop-commit] (_HOC commit_ for
+    short) is the action taken by a Hadoop OutputCommitter to finish writing
+    all objects of a job to their final locations.
+
+# Why <a href="#user-content-why" id="user-content-why">#</a>
+
+## Issues with existing OutputCommitters
+
+OutputCommitters are charged with writing a directory tree (a multi-part
+Hadoop file, for example a partitioned Parquet file) atomically.
+
+Hadoop comes with several OutputCommitter algorithms.  For S3 these try to
+bridge the gap between the Hadoop assumed semantics of a POSIX-ish
+filesystem (HDFS) and the actual semantics of an object store (S3).
+
+Of course, any Hadoop OutputCommitter that is unrelated to lakeFS can only
+ever perform HOC commits.  Any lakeFS commits on the written data will need
+to be separately managed by user code or manually.
+
+The best correct committer is probably the [magic committer][magic].  It
+works using only S3 (with its current consistency guarantees!), without
+auxiliary systems such as a small HDFS or S3Guard.  The magic committer
+uploads a directory tree by starting a multipart upload to its intended location,
+creating a completion record for that multipart upload on S3, and uploading
+the object without completing the upload.  To complete the upload, all
+completion records are processed and the objects magically appear.
+
+The [staging committer][staging] (aka "Netflix committer") is an older
+committer for S3A.  It requires a "staging" filesystem on HDFS and offers
+manual modes for overwriting outputs ("conflict resolution").
+
+These issues remain with the magic and staging committer:
+
+* Not atomic.  Multipart objects appear sequentially.  Finding a "_success"
+  object indicator is still required before a directory tree can be
+  processed.
+* The magic committer requires "magic committer" support in the FileSystem,
+  in order to write `__magic` objects to a _different_ path.  (Current
+  lakeFSFS does not contain this support.)
+* The magic committer is somewhat new:
+  > It’s also not been field tested to the extent of Netflix’s committer;
+  > consider it the least mature of the committers.
+* Documentation is still somewhat lacking.
+
+Other OutputCommitters are not as good, and _rename by copying_.  This is
```

Comment:
did you mean this? 
```suggestion
Other OutputCommitters are not as good, and use _rename by copying_.  This is
```
Also, isn't _rename by copying_ a file system property?

Reply:
True.  Reworded to split the explanation between (File) OutputCommitter and FileSystem.